File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.java
import java.util.*;
import java.io.*;

// HashMap 쓰기
class Solution {
    public int solution(String str1, String str2) {
        HashMap<String, Integer> group1 = new HashMap<>();
        HashMap<String, Integer> group2 = new HashMap<>();
        
        getHashMap(str1, group1);
        getHashMap(str2, group2);
        
        int union = 0;
        int intersection = 0;
        Set<String> unionKeySet = new HashSet<>(group1.keySet());
        unionKeySet.addAll(group2.keySet());
        
        for (String key : unionKeySet) {
            int group1_count = group1.getOrDefault(key, 0);
            int group2_count = group2.getOrDefault(key, 0);
            
            union += Math.min(group1_count, group2_count);
            intersection += Math.max(group1_count, group2_count);
        }
        
        if (union == 0 && intersection == 0) {
            return 1 * 65536;
        } else {
            return (int) (((double) union/intersection) * 65536);
        }
    }
    
    public static void getHashMap(String str, HashMap<String, Integer> group) {
         for (int i=0; i<str.length()-1; i++) {
            char left = str.charAt(i);
            char right = str.charAt(i+1);
            String word = "" + left + right;

            if (!isEnglishChar(left) || !isEnglishChar(right)) {
                continue;
            }
             
            if (group.get(word.toLowerCase()) != null) {
                group.put(word.toLowerCase(), group.get(word.toLowerCase())+1);
            } else {
                group.put(word.toLowerCase(), 1);
            }
        }
    }
    
    public static boolean isEnglishChar(char ch) {
        return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z');
    }
}

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.js
// 1. 2개씩 끊기 (str1, str2) - 특수문자 있으면 제외!!
// 2. 2개씩 끊은 값을 map에 넣기 (key는 단어, value는 count)
// 3. 양쪽 비교
function solution(str1, str2) {
    const map1 = getWord(str1);
    const map2 = getWord(str2);
    
    const set = new Set([...Object.keys(map1), ...Object.keys(map2)]);
    let count1 = 0; // 교
    let count2 = 0; // 합
    for (v of set) {
        const v1 = map1[v] ?? 0;
        const v2 = map2[v] ?? 0;
        count1 += Math.min(v1, v2);
        count2 += Math.max(v1, v2);
    }
    
    if (count1 === 0 && count2 === 0) return 65536;
    return Math.floor((count1/count2) * 65536);
}

function getWord(str) {
    const map = {};
    
    const reg = /[a-zA-Z]/;
    for (let i=0; i<str.length-1; i++) {
        let v = str[i] + str[i+1];
        v = v.toLowerCase();
        
        if (!reg.test(v[0]) || !reg.test(v[1])) continue;
        map[v] = (map[v] ?? 0) + 1;
    }
    
    return map;
}